ethereum/solidity/contracts/lib/tree/namespace/TreeHasher.sol

```diff
@@ -0,0 +1,56 @@
+// SPDX-License-Identifier: Apache-2.0
+pragma solidity ^0.8.0;
+
+import "../Constants.sol";
+import "./NamespaceNode.sol";
+
+/// @notice Get the minimum namespace.
+// solhint-disable-next-line func-visibility
+function namespaceMin(bytes8 l, bytes8 r) pure returns (bytes8) {
+    if (uint64(l) < uint64(r)) {
+        return l;
+    } else {
+        return r;
+    }
+}
+
+/// @notice Get the maximum namespace.
+// solhint-disable-next-line func-visibility
+function namespaceMax(bytes8 l, bytes8 r) pure returns (bytes8) {
+    if (uint64(l) > uint64(r)) {
+        return l;
+    } else {
+        return r;
+    }
+}
+
+/// @notice Hash a leaf node.
+/// @param minmaxNID Namespace ID.
+/// @param data Raw data of the leaf.
+// solhint-disable-next-line func-visibility
+function leafDigest(bytes8 minmaxNID, bytes memory data) pure returns (NamespaceNode memory) {
+    bytes32 digest = sha256(abi.encodePacked(Constants.LEAF_PREFIX, minmaxNID, minmaxNID, data));
+    NamespaceNode memory node = NamespaceNode(minmaxNID, minmaxNID, digest);
+    return node;
+}
+
+/// @notice Hash an internal node.
+/// @param l Left child.
+/// @param r Right child.
+// solhint-disable-next-line func-visibility
+function nodeDigest(NamespaceNode memory l, NamespaceNode memory r) pure returns (NamespaceNode memory) {
+    bytes8 min = namespaceMin(l.min, r.min);
+    bytes8 max;
+    if (l.min == Constants.PARITY_SHARE_NAMESPACE_ID) {
+        max = Constants.PARITY_SHARE_NAMESPACE_ID;
+    } else if (r.min == Constants.PARITY_SHARE_NAMESPACE_ID) {
+        max = r.max;
```

Comment:
```suggestion
    if (l.min == Constants.PARITY_SHARE_NAMESPACE_ID || r.min == Constants.PARITY_SHARE_NAMESPACE_ID) {
        max = Constants.PARITY_SHARE_NAMESPACE_ID;
```
If I understand right, `node.min <= node.max`. And since the leaves are ordered. Then, if we have one left leaf/node having a `min == PARITY_SHARE_NAMESPACE_ID`. Then, all the rest will have `{min,max} == PARITY_SHARE_NAMESPACE_ID`.
Correct me If I am missing something.

Reply:
Yes, that's the reasoning!

Reply:
Your suggested fix isn't correct thought because if the NMT contains non-parity shares, then `max` should be the maximum non-parity namespace ID. (I copied the logic in the contract directly from the specs for this reason.)